### DISCUSSION TITLE
Keycloak "Advanced Claim To Group" identity provider mapper example where incoming groups is a list/array

### DISCUSSION THREAD

**Discussion by annasarte** [keycloak/keycloak#13646] [Q&A]
Hello.
I posted this question a week ago on stackoverflow but have received 0 responses so I am hoping to have better luck here.

I am using Keycloak 18.0.2. Okta is one of my configured Identity Providers. I am using the OIDC provider (not SAML). I want to map the incoming groups claim from Okta to a user group I defined in Keycloak. I cannot find a lot of examples on how to search for a specific incoming group in the incoming list (the groups claim from Okta is a list aka array). Meaning, in Okta, if I were to do preview my token for the Okta authorization server, the token looks like this:
`
{
  ...
  "scp": [
    "openid"
  ],
  "auth_time": 1000,
  "sub": "testuser@xyz.com",
  ...
  "groups": [
    "Group1",
    "Group2",
    "Group3"
  ]...
}
`

In Keycloak, I defined a user group called something like "AsiaPacific". In the keycloak identity mapper provider detail screen, I want to say, that if the incoming group claim from Okta, which is an array of groups, contains  "Group1" then map that to the Keycloak group "AsiaPacific" but I cannot seem to make it work. What do I put in the "name" and "value" fields here, since no matter what I specify, it does not seem to work? Again, the value of the incoming groups claim is not a string but an array of strings as noted above.  I am attaching a screen shot of the Keycloak "Identity Provider Mapper" detail screen for reference.  Thanks in advance.

![keycloak_admin_idp_mapper](https://user-images.githubusercontent.com/34694111/183459990-943663c6-7fbd-4efd-a732-dc9f0b58433c.png)

**Comment by artur-baltabayev** [comment]
Hello @annasarte ,
yes you want to specify "groups" as the key (without the quotation marks).
For your case you can just use regex, which would (in the simplest case) result in the following example key/value configuration :
``
groups : .*Group1.*
``

Because the array is simply a String value in the code, Keycloak will try to match it with the regex expression.
Just make sure you have the "Regex Claim Values" option on like in the screenshot.
Hope that helps !

Greetings,
Artur

**Reply by annasarte** [reply]
Hello @artur-baltabayev.  Thank you for responding.

Yes, I had tried all that before, with several regexes and I could not get it to work.  

So I decided to investigate and downloaded the source code from github for Keycloak tag 18.0.2 which is the version I am using.  I debugged your code, and in the class `AbstractClaimMapper.java`, inside the method `public static Object getClaimValue(BrokeredIdentityContext context, String claim)` I found an issue.  You see, In my Okta Authorization Server (which can be found under Okta Admin--> Security--> API), the "groups" claim was defined to be in the "**access**" token.  And the method `getClaimValue` is expecting the groups claim I specified in your "**Advanced Claim To Group**" mapper to be in either the `VALIDATED_ACCESS_TOKEN` or the `VALIDATED_ID_TOKEN`. But since in my Authorization Server, the groups claim was in the **access** token (and during debug this was seen in the `FEDERATED_ACCESS_TOKEN_RESPONSE`), it was not finding it and hence, nothing was happening and I was not seeing groups in the returned JWT.

So once I changed my Authorization Server in Okta to have the groups claim in the **ID token** and not access token, it started to work! This was the issue all along.

Also, defining an "Advanced Claim To Group" mapper alone is not enough.  You have to also define a " Group Membership" mapper in the Client you are using within Keycloak.  Attached are screenshots (a picture is worth a thousand words right?) of all the things I had to define to ensure I got your mapper to work:

- Here is the updated "groups" Claims definition in my Okta Authorization Server to reflect that the claim is for the "**ID**" token.

![1AnnaIAMAuthServerInOkta_Claims](https://user-images.githubusercontent.com/34694111/184058840-8e74b499-515d-4259-b8f4-a80b05df5760.png)

- Here is the "Advanced Claim To Group Mapper" (I wanted to match the entire group name hence my regex is defined this way):

![Screen Shot 2022-08-10 at 8 33 18 PM](https://user-images.githubusercontent.com/34694111/184060761-a714556d-fd5b-4137-b752-484be39ef936.png)

- And here is the "Group Membership" mapper for my client:

![GroupMembershipClientMapper](https://user-images.githubusercontent.com/34694111/184059275-78769dd0-9650-46be-8f5c-256b95e16087.png)

- Finally, the end result:  The generated token that was returned from Keycloak when I logged in via the Federated Okta IdP shows the mapping to my Keycloak group in the resulting groups claim in the JWT (this screen is a simple app we wrote to show the token):

![JWT_showsGroups](https://user-images.githubusercontent.com/34694111/184060436-39babf20-fb5b-4679-9426-6d0950d40a58.png)

**Reply by sdandois** [reply]
This worked for me.

**Reply by lerian-studio** [reply]
I have a question: Do I just create a single group in this integration between Okta and Keycloak?

For example, if you input  in the menu CLAIMS in Keycloak, in values more than a single group, How to do directly the group do you add for keycloak?

**Reply by annasarte** [reply]
Hello @lerian-studio . Since I opened this question (and answered it), I also found another way to map groupS (plural) from my Identity Provider (Okta) down to Keycloak. 

- in Okta, you can assign the user as many groups as you wish (group assignments tab in Okta for the user in question)
- in Keycloak, it is 2 parts:
- first create a mapper in your Identity Provider definition (Identity Provider-> Provider details -> Mapper tab, like this:

![idp_group_mapper](https://github.com/keycloak/keycloak/assets/34694111/e9b44f9d-651b-490f-89f0-73293d90c8b8)

- second, in your Keycloak client (that is a browser/auth code client), go to clients --> client details --> Dedicated scopes -> , define a mapper within the dedicated scope like this: 
![client_dedicated_scope_group_mapper](https://github.com/keycloak/keycloak/assets/34694111/2d538170-9722-48ab-b6da-170ca6db5885)

- Make sure "Multivalued" is set to true so that you will get an array of strings back for your group.
When your Okta user logs in to the application using the Keycloak client (the browser client you are using for you application), the token generated will then have the groups in the JWT.

Hope that helps!
